### ISSUE TITLE
Block array mapping in Chunks is not correct

### ISSUE THREAD

**Issue by james0x0A** [LambdaSix/InfiniMap#12]
The current mapping is `_blocks[blockX + _chunkWidth * (blockY + _chunkDepth * blockZ)`

```
blockX + _chunkWidth * (blockY + _chunkDepth * blockZ) 
 = X + Y_SIZE * ( Y + Z_SIZE * Z) 
 = X + ( Y  * Y_SIZE ) + ( Z * Y SIZE * Z_SIZE )
```

This does not map correctly unless  the chunk is a perfect cube or square.

To map  a general cuboid space, instead use :
```
X + Y * X_SIZE + Z * X_SIZE * Y_SIZE
= X + ( Y  +  Z   *  Y_SIZE)  *  X_SIZE
= blockX +  (blockY +  blockZ  * _chunkWidth ) * _chunkHeight  

_blocks[blockX +  (blockY +  blockZ  * _chunkWidth ) * _chunkHeight]
```

**Comment by LambdaSix** [comment]
The implicit assumption was always that a given chunk would be a cube/square of the same dimensions, but this seems like a general improvement that can be made.

**Comment by james0x0A** [comment]
Just a note, if any change is made to swap  usage of the width and height variable names,  then the general mapping would  be  `blockX + (blockY + blockZ * _chunkWidth) * _chunkHeight`

**Comment by LambdaSix** [comment]
It's a breaking change. So humph, while I was tidying up the Width/Height ordering, I also figured I should take the opportunity to make another breaking change with regards the Write/Read funcs, they currently emit/accept an IEnumerable<T> as the chunk data, but since now the array layout is different, it seems prudent to make the Chunk<U> class public and emit/consume Chunk<U>'s in the Read/Write functions so that we avoid requiring the API consumer to know the internal format.

If serializing the sequence as given, then reading it back exactly the same then it should be fine with the IEnumerable<T>, but when generating chunks in-memory the user has to be aware of the internal storage format. At least this way they can continue to treat the map without caring about how it's internally stored, even if in-memory generation becomes chunk-aware on the user level.

For serialization, I think continuing to provide an IEnumerable<T> is fine, as that's pushing from memory to disk, so the user can just write the format directly.

**Comment by LambdaSix** [comment]
Should be resolved with PR #14
